protocols/observed/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Connection upgrade to allow retrieving the externally visible address (as dialer) or
+//! to report the externally visible address (as listener).
+
+extern crate bytes;
+extern crate futures;
+extern crate libp2p_core;
+extern crate tokio_codec;
+extern crate tokio_io;
+extern crate unsigned_varint;
+
+use bytes::Bytes;
+use futures::{future, prelude::*};
+use libp2p_core::{ConnectionUpgrade, Endpoint, Multiaddr};
+use std::{io, iter};
+use tokio_codec::Framed;
+use tokio_io::{AsyncRead, AsyncWrite};
+use unsigned_varint::codec::UviBytes;
+
+/// The output, this connection upgrade produces.
+pub enum Output<C> {
+    /// As `Dialer`, we get our own externally observed address.
+    ObservedAddr(Multiaddr),
+    /// As `Listener`, we return the underlying resource.
+    Resource(Framed<C, UviBytes>)
+}
+
+/// The connection upgrade type to retrieve or report externally visible addresses.
+pub struct Observed {}
+
+impl Observed {
+    pub fn new() -> Self {
+        Observed {}
+    }
+}
+
+impl<C, Maf> ConnectionUpgrade<C, Maf> for Observed
+where
+    C: AsyncRead + AsyncWrite + Send + 'static,
+    Maf: Future<Item=Multiaddr, Error=io::Error> + Send + 'static
+{
+    type NamesIter = iter::Once<(Bytes, Self::UpgradeIdentifier)>;
+    type UpgradeIdentifier = ();
+    type Output = Output<C>;
+    type MultiaddrFuture = Box<dyn Future<Item=Multiaddr, Error=io::Error> + Send>;
+    type Future = Box<dyn Future<Item=(Self::Output, Self::MultiaddrFuture), Error=io::Error> + Send>;
+
+    fn protocol_names(&self) -> Self::NamesIter {
+        iter::once((Bytes::from("/observed-address/0.1.0"), ()))
```

Comment:
Since this is "experimental", I think we should have a "non-standard" protocol name.

```suggestion
        iter::once((Bytes::from("/paritytech/observed-address/0.1.0"), ()))
```